caldav/match.go

```diff
@@ -0,0 +1,224 @@
+package caldav
+
+import (
+	"time"
+
+	"github.com/emersion/go-ical"
+)
+
+// Filter returns the filtered list of calendar objects matching the provided query.
+// A nil query will return the full list of calendar objects.
+func Filter(query *CalendarQuery, cos []CalendarObject) ([]CalendarObject, error) {
+	if query == nil {
+		// FIXME: should we always return a copy of the provided slice?
+		return cos, nil
+	}
+
+	out := make([]CalendarObject, 0)
+	for _, co := range cos {
+		ok, err := Match(query.CompFilter, &co)
+		if err != nil {
+			return nil, err
+		}
+		if !ok {
+			continue
+		}
+
+		// TODO properties are not currently filtered even if requested
+		out = append(out, co)
+	}
+	return out, nil
+}
+
+// Match reports whether the provided CalendarObject matches the query.
+func Match(query CompFilter, co *CalendarObject) (matched bool, err error) {
+	if co.Data == nil || co.Data.Component == nil {
+		// TODO is this a thing? Should this match if comp.IsNotDefined?
+		return false, nil
+	}
+	comp := co.Data.Component
+
+	if query.Name != comp.Name {
```

Comment:
According to https://datatracker.ietf.org/doc/html/rfc4791#section-9.7.1, `query.Name` may match a child of the calendar object. `VCALENDAR` is the root component, but the request might match a child VEVENT component.

Reply:
Hm, in that VEVENT case it seems like the child comp-filter must be enclosed in a parent VCALENDAR comp-filter, so should be fine.